src/interval_sets.jl

```diff
@@ -481,95 +481,55 @@ find_intersections(x, y) = find_intersections(vcat(x), vcat(y))
     tracking = endpoint_tracking(x, y)
     lt = intersection_isless_fn(tracking)
-    x_endpoints = unbunch(enumerate(x), tracking; lt)
-    y_endpoints = unbunch(enumerate(y), tracking; lt)
-    result = [Vector{Int}() for _ in 1:length(x)]
-
-    return find_intersections_helper!(result, x_endpoints, y_endpoints, lt)
-end
-
-function find_intersections_helper!(result, x, y, lt)
-    active_xs = Set{Int}()
-    active_ys = Set{Int}()
-    while !isempty(x)
-        xᵢ, yᵢ = first_endpoint(x), first_endpoint(y)
-        x_less = lt(xᵢ, yᵢ)
-        y_less = lt(yᵢ, xᵢ)
-
-        if !y_less
-            if isleft(xᵢ)
-                push!(active_xs, first(first(x)))
-            else
-                delete!(active_xs, first(first(x)))
-            end
-            x = @view x[2:end]
-        end
-
-        if !x_less
-            if isleft(yᵢ)
-                push!(active_ys, first(first(y)))
-            else
-                delete!(active_ys, first(first(y)))
-            end
-            y = @view y[2:end]
-        end
-
-        for i in active_xs
-            append!(result[i], active_ys)
-        end
-    end
+    results = Vector{Vector{Int}}(undef, length(x))
 
-    return unique!.(result)
+    return find_intersections_helper!(results, x, y, lt)
 end
 
-function find_intersections(
-    x::AbstractVector{<:AbstractInterval{T1,Closed,Closed}},
-    y::AbstractVector{<:AbstractInterval{T2,Closed,Closed}},
-) where {T1,T2}
+function find_intersections_helper!(results, x, y, lt)
     # Strategy:
     # two binary searches per interval `I` in `x`
     # * identify the set of intervals in `y` that start during-or-after `I`
     # * identify the set of intervals in `y` that stop before-or-during `I`
     # * intersect them
-    starts = first.(y)
-    starts_perm = sortperm(starts)
-    starts_sorted = starts[starts_perm]
+    lefts = LeftEndpoint.(y)
+    lefts_order = sortperm(lefts; lt)
+    lefts_sorted = lefts[lefts_order]
 
     # Sneaky performance optimization (makes a huge difference!)
-    # Rather than sorting `stops` relative to `y`, we sort it relative to `starts`.
-    # This allows us to work in the `starts` frame of reference until the very end.
-    # In particular, when we intersect the sets of intervals obtained from starts and from stops,
-    # the `starts` set can be kept as a `UnitRange`, making the intersection *much* faster.
-    stops = last.(y[starts_perm])
-    stops_perm = sortperm(stops)
-    stops_sorted = stops[stops_perm]
-    len = length(stops_sorted)
+    # Rather than sorting `rights` relative to `y`, we sort it relative to `lefts`.
+    # This allows us to work in the `lefts` frame of reference until the very end.
+    # In particular, when we intersect the sets of intervals obtained from lefts and from rights,
+    # the `lefts` set can be kept as a `UnitRange`, making the intersection *much* faster.
+    rights = RightEndpoint.(y)
+    rights_order = sortperm(rights[lefts_order]; lt)
```

Comment:
Oversight on my part that results in `rights_sorted` not actually being sorted.
```suggestion
    rights = RightEndpoint.(y)[lefts_order]
    rights_order = sortperm(rights; lt)
```